### PR TITLE
Loading Service Instances in BTP Manager UI Backend - API

### PR DESCRIPTION
**Pull request by ralikio**
**Description**

The following PR connects methods previously introduced in https://github.com/kyma-project/btp-manager/pull/742 and exposes real data on endpoints `/api/service-instances`.

Changes proposed in this pull request:

- list instances based on previously written service offerings in BTP Manager UI backend.
**Related issue(s)**

#442

### CODE REVIEW

**Comment by ralikio** [comment]
Current response from the API: 
```
{
    "numItems": 1,
    "items": [
        {
            "name": "auditlog-management",
            "namespace": "",
            "id": "4e27f92d-872a-4410-9076-37fe4ac75d4c",
            "description": "",
            "catalogName": "",
            "metadata": {
                "imageUrl": "",
                "displayName": "",
                "documentationUrl": "",
                "supportUrl": ""
            }
        }
    ]
}
```

**Comment by ralikio** [comment]
API changes were picked up by UI:
image

File: internal/api/vm/vm.go [hunk 1]
```diff
@@ -30,12 +30,18 @@ type ServiceOfferingMetadata struct {
 }
 
 type ServiceInstances struct {
-	Items []ServiceInstance `json:"items"`
+	NumItems int               `json:"numItems"`
```
Comment (jaroslaw-pieszka) [inline internal/api/vm/vm.go:33]: Why do you change the struct and introduce redundant field?

File: internal/service-manager/types/service_instance.go [hunk 2]
```diff
@@ -2,6 +2,9 @@ package types
 
 import "encoding/json"
 
+// Refs
+// https://github.com/SAP/sap-btp-service-operator/blob/main/client/sm/types/service_instance.go
+// https://github.com/Peripli/service-manager/blob/master/pkg/types/service_instance.go
```
Comment (jaroslaw-pieszka) [inline internal/service-manager/types/service_instance.go:7]: Below NumItems is not added.

File: internal/api/vm/vm.go (hunk 1 above)
Comment (ralikio) [reply internal/api/vm/vm.go:33]: Done for consistency with what has been done for service offerings.

File: internal/service-manager/types/service_instance.go (hunk 2 above)
Comment (ralikio) [reply internal/service-manager/types/service_instance.go:7]: This is a model for the Service Manager. I guess that `NumItems` was applicable just in VM structs (View Model), maybe, to show count in the UI.

File: internal/api/vm/vm.go (hunk 1 above)
Comment (jaroslaw-pieszka) [reply internal/api/vm/vm.go:33]: Needs further discussion since secrets do not contain NumItems and we still have inconsistency.

File: internal/api/api.go [hunk 3]
```diff
@@ -46,19 +46,22 @@ func NewAPI(cfg Config, serviceManager *servicemanager.Client, secretProvider *c
...
-
+	router := http.NewServeMux()
+	a.AttachRoutes(router)
+	a.server.Handler = router
 	log.Fatal(a.server.ListenAndServe())
 }
 
+func (a *API) AttachRoutes(router *http.ServeMux) {
+	router.HandleFunc("GET /api/secrets", a.ListSecrets)
+	router.HandleFunc("GET /api/service-instances", a.ListServiceInstances)
```
Comment (jaroslaw-pieszka) [inline internal/api/api.go:57]: Unlike for service-offerings we do not allow to specify namespace and name. Why is that? Should it be consistent or not.

File: internal/api/api_test.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+							ID:          "2",
+							Name:        "service-2",
+							Description: "",
+						},
+					},
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:60]: Kind of anti-YAGNI.

File: internal/api/api_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+			require.NoError(t, err)
+			validateJSON(t, got, tt.file)
+		})
+	}
+}
+
+func validateJSON(t *testing.T, got []byte, file string) {
+	var prettyWant bytes.Buffer
+	want := readJsonFile(t, file)
+	if len(want) > 0 {
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:93]: Code duplication

File: internal/api/api_test.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+			got, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+			validateJSON(t, got, tt.file)
+		})
+	}
+}
+
+func validateJSON(t *testing.T, got []byte, file string) {
+	var prettyWant bytes.Buffer
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:91]: Misleading name.

File: internal/api/api_test.go [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+	var prettyGot bytes.Buffer
+	if len(got) > 0 {
+		err := json.Indent(&prettyGot, got, "", "  ")
+		if err != nil {
+			t.Error(err)
+			t.Fail()
+		}
+	}
+	if !assert.JSONEq(t, prettyGot.String(), prettyWant.String()) {
+		t.Errorf("%v Schema() = \n######### GOT ###########%v\n######### ENDGOT ########, want \n##### WANT #####%v\n##### ENDWANT #####", file, prettyGot.String(), prettyWant.String())
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:110]: More commonly `expected` is used.

File: internal/api/api_test.go (hunk 4 above)
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:60]: No test cases for unhappy path.

File: internal/api/vm/converters.go [hunk 8]
```diff
@@ -48,6 +48,23 @@ func ToServiceOfferingsVM(offerings *types.ServiceOfferings) ServiceOfferings {
 	return toReturn
 }
 
+func ToServiceInstancesVM(instances *types.ServiceInstances) ServiceInstances {
```
Comment (jaroslaw-pieszka) [inline internal/api/vm/converters.go:51]: Is this function somehow tested using unit test approach?
Testing if NumItems == len(Items) for corner cases?

File: internal/api/api_test.go [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+			got, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+			validateJSON(t, got, tt.file)
+		})
+	}
+}
+
+func validateJSON(t *testing.T, got []byte, file string) {
+	var prettyWant bytes.Buffer
+	want := readJsonFile(t, file)
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:92]: `expected`?

File: internal/api/api_test.go [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	if !assert.JSONEq(t, prettyGot.String(), prettyWant.String()) {
+		t.Errorf("%v Schema() = \n######### GOT ###########%v\n######### ENDGOT ########, want \n##### WANT #####%v\n##### ENDWANT #####", file, prettyGot.String(), prettyWant.String())
+	}
+}
+
+func readJsonFile(t *testing.T, file string) string {
+	t.Helper()
+
+	filename := path.Join("testdata", file)
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:117]: magic value

**Comment by szwedm** [comment]
Hey @ralikio, I would like to discuss the introduced changes, applying on hold label until then.

File: internal/api/api_test.go (hunk 6 above)
Comment (ralikio) [reply internal/api/api_test.go]: Corrected to prettyExpected if that is what you mean.

File: internal/api/api_test.go (hunk 4 above)
Comment (ralikio) [reply internal/api/api_test.go:60]: This is simple API and tests are about to be refactored. Please propose test cases that you would like to see.

**Comment by ralikio** [comment]
List Instances were implemented in https://github.com/kyma-project/btp-manager/pull/735.

File: internal/api/api_test.go (hunk 4 above)
Comment (jaroslaw-pieszka) [reply internal/api/api_test.go:59]: `ListInstances` has two paths of execution. Happy (returning response and status 200) and unhappy (returning error response and 503. 
Test for unhappy flow would detect e.g. change in error status, and error handling in UI supposedly relies on some assumptions about statuses returned from API.

File: internal/api/api_test.go (hunk 10 above)
Comment (jaroslaw-pieszka) [reply internal/api/api_test.go:114]: https://refactoring.guru/replace-magic-number-with-symbolic-constant

Comment (jaroslaw-pieszka) [reply internal/api/api_test.go:114]: I could agree that this is nit picking but even if extracting this literal as string constant seems to be overkill there are lots of developers preferring this approach. Mostly in Java world.

File: internal/api/vm/vm.go (hunk 1 above)
Comment (ralikio) [reply internal/api/vm/vm.go]: Outdated. I am not changing this struct in my PR anymore.

File: internal/api/api.go (hunk 3 above)
Comment (ralikio) [reply internal/api/api.go:62]: I am not changing the path in this PR.

File: internal/api/api_test.go (hunk 4 above)
Comment (ralikio) [reply internal/api/api_test.go:60]: Added 503 case in https://github.com/kyma-project/btp-manager/pull/746/commits/ca427c8fcabedcaafb278c93ffa44890afd900cc

File: internal/api/api_test.go (hunk 10 above)
Comment (ralikio) [reply internal/api/api_test.go]: Addressed in https://github.com/kyma-project/btp-manager/pull/746/commits/d736bf605e297ac2e02e46df9c3fd15c5c903223

File: internal/api/vm/converters.go (hunk 8 above)
Comment (ralikio) [reply internal/api/vm/converters.go]: Not changing NumItems in the pr after its rebase.

File: internal/api/api_test.go [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	expected := readJsonFile(t, file)
+	prettyExpected := indent([]byte(expected), t)
+	prettyGot := indent(got, t)
+
+	if !assert.JSONEq(t, prettyGot.String(), prettyExpected.String()) {
+		t.Errorf("%v Schema() = \n######### GOT ###########%v\n######### ENDGOT ########, expected \n##### EXPECTED #####%v\n##### ENDEXPECTED #####", file, prettyGot.String(), prettyExpected.String())
+	}
+}
+
+func indent(expected []byte, t *testing.T) *bytes.Buffer {
```
Comment (jaroslaw-pieszka) [inline internal/api/api_test.go:123]: Seems redundant. Is there really problem with `json.indent` when called with empty source/

Comment (ralikio) [reply internal/api/api_test.go:123]: When called with empty "expected" param we get the following error:
```Expected value ('') is not valid json. JSON parsing error: 'unexpected end of JSON input'```

Comment (ralikio) [reply internal/api/api_test.go:123]: Addressed in https://github.com/kyma-project/btp-manager/pull/746/commits/5681c59606b69b15a9724899839b42e26a36c787

**Review by jaroslaw-pieszka — APPROVED** [review]
